File: KHRiskFactorManager.m
//
//  KHRiskFactorManager.m
//  KHHealthCareProj
//
//  Created by Tyler Lu on 7/22/16.
//  Copyright © 2016 Ryan Lu. All rights reserved.
//

#import "KHRiskFactorManager.h"
#import "KHHealthCareProj-Swift.h"
#import "KHRiskFactor.h"

@implementation KHRiskFactorManager

+ (id)sharedManager {
	static KHRiskFactorManager *sharedMyManager = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		sharedMyManager = [[self alloc] init];
	});
	return sharedMyManager;
}

- (void)downloadAllRiskFactors {
	
	//declare the mutable copies of the risk factors
	NSMutableArray *allRiskFactorsCopy = [[NSMutableArray alloc] init];
	NSMutableArray *ageRiskFactorsCopy = [[NSMutableArray alloc] init];
	NSMutableArray *EFLRiskFactorsCopy = [[NSMutableArray alloc] init];
	NSMutableArray *MCRiskFactorsCopy = [[NSMutableArray alloc] init];
	
	NSMutableArray *generalRiskFactorsCopy = [[NSMutableArray alloc] init];
	NSMutableArray *cancerRiskFactorsCopy = [[NSMutableArray alloc] init];
	NSMutableArray *vaccineRiskFactorsCopy = [[NSMutableArray alloc] init];
	
	FIRDatabaseReference *ref = [[FIRDatabase database] reference];
	
	
	[ref observeSingleEventOfType:FIRDataEventTypeValue withBlock:^(FIRDataSnapshot * _Nonnull snapshot) {
		NSDictionary *base = snapshot.value;

		NSDictionary *rfDict = [base objectForKey: @"riskfactors-general"];
		NSDictionary *cateDict = [base objectForKey:@"rf-categories"];
		NSDictionary *subCateDict = [base objectForKey:@"rf-sub-categories"];
		NSDictionary *gsDict = [base objectForKey:@"general screening"];
		NSDictionary *gsCateDict = [base objectForKey:@"general screening-categories"];
		
//		NSLog(@"%@", rfDict);
//		NSLog(@"%@", cateDict);
//		NSLog(@"%@", subCateDict);
//		NSLog(@"%@", gsDict);
		
		//Parse the risk factors 
		[rfDict enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
			NSDictionary *rfValueDict = [rfDict objectForKey:key];
			NSLog(@"value dict: %@", [rfValueDict description]);
			
			NSString *name = [rfValueDict objectForKey:@"name"];
			NSString *ID = [rfValueDict objectForKey:@"id"];
			NSString *categoryIndex = rfValueDict[@"category"];
			
			__block NSMutableString *category;
			
			//Find the category names for this risk factor
			[cateDict enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
                
				if ([categoryIndex isEqualToString: (NSString *)key]) {
					category = [NSMutableString stringWithString: cateDict[key]] ;
					
				}
			}];
			
			
			//Find the sub-category name
			NSString *subcatIndex = rfValueDict[@"sub-category"];
			__block NSMutableString *subCate;
			
			[subCateDict enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
				if ([subcatIndex isEqualToString:(NSString *)key]) {
					NSDictionary *subCateValueDict = subCateDict[key];
					subCate = subCateValueDict[@"name"];
				}
			}];
			
//			NSLog(@"%@   %@", category, subCate);
			
			//========================== general list =======================
			
			//TODO: Fix parser function
			NSDictionary *general_list = [NSDictionary dictionary];
			
			if (rfValueDict[@"general-list"] != nil) {
				
				NSDictionary *list = rfValueDict[@"general-list"];
				
				general_list = [NSDictionary dictionaryWithDictionary:[self parseListDictWithDict:list  andDefDict:gsDict]];
			}else {
				NSLog(@"empty general list for RF");
			}
			NSLog(@"general list for rf: %@ list: %@", name, general_list);
			
			
			//======================= vaccine list ========================
			NSDictionary *vaccine_list;
			NSDictionary *list = rfValueDict[@"vaccine-list"];
			
			if (list != nil) {
//				vaccine_list = [self parseListDictWithDict:list andListDefDict:gsDict];
			}
			NSLog(@"Vaccine List: %@", vaccine_list);
			
			
			//======================== cancer list ===========================
			NSDictionary *cancer_list;
			list = rfValueDict[@"cancer-list"];
			if (list != nil) {
//				cancer_list = [self parseListDictWithDict:list andListDefDict:gsDict];
			}
			NSLog(@"Cancer List: %@", vaccine_list);
			
			
			//=================== init risk factor object ====================
			KHRiskFactor* riskfactor = [[KHRiskFactor alloc] initWithName:name category:category ID:ID subCategory:subCate generalList:general_list vaccineList:vaccine_list cancerList:cancer_list];
			
			//All Riskfactors mutable copy
			[allRiskFactorsCopy addObject:riskfactor];
			
			if ([category isEqualToString:@"Age"])										//Age riskfactor mutable copy
				[ageRiskFactorsCopy addObject:riskfactor];
			else if ([category isEqualToString:@"Medical Condition"])					//MD riskfactor mutable copy
				[MCRiskFactorsCopy addObject:riskfactor];
			else if ([category isEqualToString:@"Ethnicity, Family, Lifestyle"])		//EFL risk factor mutable copy
				[EFLRiskFactorsCopy addObject:riskfactor];
			
			if (riskfactor.generalList != nil)
				[generalRiskFactorsCopy addObject:riskfactor];
			if (riskfactor.vaccineList != nil)
				[vaccineRiskFactorsCopy addObject:riskfactor];
			if (riskfactor.cancerList != nil)
				[cancerRiskFactorsCopy addObject:riskfactor];
			
//			NSLog(@"%@", riskfactor.name);
//			NSLog(@"%lu %lu", allRiskFactorsCopy.count, rfDict.count);
			
			//after all the risk factors are parsed save them into riskfactors
			if (allRiskFactorsCopy.count >= rfDict.count) {
				self.allRiskFactors = [NSArray arrayWithArray:allRiskFactorsCopy];
				self.EFLRiskFactors = [NSArray arrayWithArray:EFLRiskFactorsCopy];
				self.MedicalCondRiskFactors = [NSArray arrayWithArray:MCRiskFactorsCopy];
				
				self.generalRiskFactors = [NSArray arrayWithArray:generalRiskFactorsCopy];
				self.vaccineRiskFactors = [NSArray arrayWithArray:vaccineRiskFactorsCopy];
				self.cancerRiskFactors = [NSArray arrayWithArray:cancerRiskFactorsCopy];
				
//				NSLog(@"%lu", self.EFLRiskFactors.count);
			}
		}];
        NSLog(@"all riskf factors: %@", allRiskFactorsCopy);
        for (KHRiskFactor *rf in allRiskFactorsCopy) {
            NSLog(@"------------------------------");
            NSLog(@"riskfactor name: %@", rf.name);
            NSLog(@"category: %@", rf.category);
            
        }
	}];
	
}
/* 
 * Helper function to parse the list
 */
- (NSDictionary*) parseListDictWithDict:(NSDictionary*)valueDict andDefDict:(NSDictionary *)defDict {
	
	NSLog(@"%@  \n\n %@", valueDict, defDict);
	
	NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
	
	
	for (NSString *key in valueDict) {
		NSMutableDictionary *screeningDict =[NSMutableDictionary dictionaryWithDictionary:defDict[key]];
		NSLog(@"%@", screeningDict);
		if (screeningDict == nil || screeningDict[@"name"] == nil) {
			continue;
		}

		NSLog(@"%@",[valueDict[key] objectForKey:@"value"]);
		[screeningDict setObject:[valueDict[key] objectForKey:@"value"] forKey:@"value"];
		
		NSString *screeningName = screeningDict[@"name"];
		
		[result setObject:screeningDict forKey:screeningName];
	}
	
	
	NSLog(@"%@",result);
	return result;
}



@end

File: KHRiskFactorManager.swift
//
//  KHRiskFactorModel.swift
//  KHHealthCareProj
//
//  Created by Ryan Lu on 6/14/16.
//  Copyright © 2016 Ryan Lu. All rights reserved.
//
// This is the model that holds all the risk factors

import Foundation
import FirebaseDatabase


class KHRiskFactorManagerSwift: NSObject {
    static let sharedManager = KHRiskFactorManagerSwift()
	
	//MARK: - Properties
	private var _allRiskFactors:[KHRiskFactor] = []
	private var _ageRiskFactors:[KHRiskFactor] = []
	private var _MDRiskFactors:[KHRiskFactor] = []
	private var _EFLRiskFactors:[KHRiskFactor] = []
	
	//TODO: Parse and save categories
	private var _RFCategories:Dictionary<String,String> = [:]
	private var _RFSubCategories:Dictionary<String,String> = [:]
	
	//MARK: - GETTER & SETTER
	var allRiskFactors:[KHRiskFactor] {
		get {return _allRiskFactors}
	}
	var ageRiskFactors:[KHRiskFactor] {
		get {return _ageRiskFactors}
	}
	var MDRiskFactors:[KHRiskFactor]  {
		get {return _MDRiskFactors}
	}
	var EFLRiskFactors:[KHRiskFactor] {
		get {
			return _EFLRiskFactors
		}
	}
	
	//MARK: - INIT
	override init() {
		
	}
	
	// MARK: - DOWNLOAD DATA FUNCS
	func downloadAllRiskFactors( completion:(completed:Bool) -> Void){
		let ref = FIRService.sharedService.refBase
	
		ref.observeSingleEventOfType(.Value) { (snapshot:FIRDataSnapshot) in
			if let base = snapshot.value as? Dictionary<String, AnyObject> {
                
                print("base: \(base)")
                let rfDict = base["riskfactors-general"]
//                print("rfdict: \(rfDict)")
                let cateDict = base["rf-categories"]
                let subCateDict = base["rf-sub-categories"]
                let gsDict = base["general screening"] as! [String:AnyObject]
                let gsCateDict = base["general screening-categories"]
                
                for (rfKey,rfValue) in rfDict as! [String:AnyObject] {
					
//                    print("riskfactor key: \(rfKey) , value: \(rfValue)")
					let rfValueDict = rfValue as! [String:AnyObject]
                    
                    
                    //get name
					
					let name = rfValueDict["name"] as! String
                    
                    //get id
					let id = rfValueDict["id"] as! String
                    
                    //get category
					let categoryIndex = rfValueDict["category"] as! String
                    var category:String = ""
                    for (cateKey, cateValue) in cateDict as! [String:String] {
//                        print("cate key and cate value: \(cateKey) , \(cateValue)")
                        if (categoryIndex == cateKey ) {
                            category = cateValue
                        }
                    }
//                    print("done getting category")
                    
                    //get sub-category
					let subcatIndex = rfValueDict["sub-category"] as! String
                    var subCate:String = ""
//                    print("here: \(subCateDict)")
                    for (subCateKey, subCateValue) in subCateDict as! [String:AnyObject] {
//                        print("subcate key and subcate value: \(subCateKey), \(subCateValue)")
                        
                        if (subcatIndex == subCateKey ) {
                            let subCateValueDict = subCateValue as! [String:String]
                            subCate = subCateValueDict["name"]!
                            
                        }
                    }
//					print("done getting subcategory")
                    
                    //get general list
					var general_list:[String:String]?
					if let list = rfValueDict["general-list"] as? [String:AnyObject] {
						general_list = self.parseListDict(list, listDefDict:gsDict)
					}
                    print("general list for rf: \(name), list: \(general_list)")
                    
					
                    //get vaccine list
					var vaccine_list:[String:String]?
					if let list = rfValueDict["vaccine-list"] as? [String:AnyObject] {
						vaccine_list = self.parseListDict(list, listDefDict: gsDict)
					}
					
                    //get cancer list
                    
                    
                    
                    //declare riskfactor
//					let riskfactor = KHRiskFactor(name: name, category: category, id: id, subcategory: subCate, generalList: general_list, vaccineList: vaccine_list, cancerList: nil)
                    
                    
                    //add riskfactor to list
//					self._allRiskFactors.append(riskfactor)
					
	
                
				}
                
                
				
//                print("all riskfactors added: \(self._allRiskFactors.count)")
//                print("category one (efl): \(self._EFLRiskFactors.count)")
//                for rf in self._allRiskFactors {
//                    print("rf name: \(rf.name)")
//                    print("rf category: \(rf.category)")
//                    print("rf id: \(rf.id)")
//                    print("rf sub-category: \(rf.subcategory)")
//                    print("general dict: \(rf.generalList)")
//                    print("vaccine dict: \(rf.vaccineList)")
//                    print("-------------------------------")
//                }
//                for i in 1 ..< self._EFLRiskFactors.count {
//                    self._EFLRiskFactors
//                }
				completion(completed: true)
                
				
			}
		}
	}
	
	//MARK:  - Helper Functions
    private func parseListDict(valueDict:[String:AnyObject], listDefDict:[String:AnyObject]) -> [String:String] {
		var listDict:[String:String] = [:]
		/* var name
        var category
        var info
        var id =
        var value = */
        
        
		for (key,value) in valueDict {
            //for key index for definition
            /* for (defKey, defValue) in listDefDict {
//                let defVaueDict =
                if key == defKey {
                    let name = defValue["name"] as! String
                    let category = defValue["category"] as! String
                    let info = defValue["info"] as! String
                    let id = defValue["id"] as! String
                    let screening = KHScreening(name: name, category: category, id: id, info: info)
                    listDict[
                }
            } */
            
            
            //get value for that index
			guard let dict = value as? [String:String], let val = dict["value"]  else {
				continue
			}
			
            
            
            
			listDict[key] = val
		}

		return listDict
	}

	
	
}